skbio/metadata/_metadata.py

```diff
@@ -0,0 +1,1337 @@
+    def _validate_index(cls, index, *, axis):
+        if axis == 'id':
+            label = 'ID'
+        elif axis == 'column':
+            label = 'column name'
+        else:
+            raise NotImplementedError
+
+        for value in index:
+            if not isinstance(value, str):
+                raise TypeError(
+                    "Detected non-string metadata %s of type %r: %r" %
+                    (label, type(value), value))
+
+            if not value:
+                raise ValueError(
+                    "Detected empty metadata %s. %ss must consist of at least "
+                    "one character." % (label, label))
+
+            if axis == 'id' and value.startswith('#'):
+                raise ValueError(
+                    "Detected metadata %s that begins with a pound sign "
+                    "(#): %r" % (label, value))
+
+            if is_id_header(value):
+                raise ValueError(
+                    "Detected metadata %s %r that conflicts with a name "
+                    "reserved for the ID header. Reserved ID headers:\n\n%s" %
+                    (label, value, FORMATTED_ID_HEADERS))
+
+        if len(index) != len(set(index)):
+            duplicates = find_duplicates(index)
+            raise ValueError(
+                "Metadata %ss must be unique. The following %ss are "
+                "duplicated: %s" %
+                (label, label, ', '.join(repr(e) for e in sorted(duplicates))))
+
+    @classmethod
+    def _filter_ids_helper(cls, df_or_series, ids, ids_to_keep):
+        # `ids_to_keep` can be any iterable, so turn it into a list so that it
+        # can be iterated over multiple times below (and length-checked).
+        ids_to_keep = list(ids_to_keep)
+
+        if len(ids_to_keep) == 0:
+            raise ValueError("`ids_to_keep` must contain at least one ID.")
+
+        duplicates = find_duplicates(ids_to_keep)
+        if duplicates:
+            raise ValueError(
+                "`ids_to_keep` must contain unique IDs. The following IDs are "
+                "duplicated: %s" %
+                (', '.join(repr(e) for e in sorted(duplicates))))
+
+        ids_to_keep = set(ids_to_keep)
+        missing_ids = ids_to_keep - ids
+        if missing_ids:
+            raise ValueError(
+                "The following IDs are not present in the metadata: %s"
+                % (', '.join(repr(e) for e in sorted(missing_ids))))
+
+        # While preserving order, get rid of any IDs not contained in
+        # `ids_to_keep`.
+        ids_to_discard = ids - ids_to_keep
+        return df_or_series.drop(labels=ids_to_discard, axis='index',
+                                 inplace=False, errors='raise')
+
+    def save(self, filepath, ext=None):
+        """Save a TSV metadata file.
+
+        The TSV metadata file format is described at https://docs.qiime2.org in
+        the Metadata Tutorial.
+
+        The file will always include the ``#q2:types`` directive in order to
```

Comment:
We should be backward compatible with `q2:types` but this may need to be parameterized so that it reflects skbio by default

Reply:
Sorry, I meant that `q2:types` doesn't make sense as a name for scikit-bio, and instead should perhaps default to `sk:types` or something to that effect. I don't know if QIIME 2, outside of this object, expect details about the exact label used. If so then it should be controllable but the calling library

Reply:
oh ... yes that is a good point.  If we ditch `q2:types`, that'll definitely incur breaking changes...

Currently, we are planning on fleshing out the `q2:types` functionality into the `DataDictionary` object.
If we are going to have breaking changes anyways, does it make sense to implement this redundant functionality?

CC @gregcaporaso